From 93a4b6a0a67a211a2e69b198b0e6a98a941d64ce Mon Sep 17 00:00:00 2001
From: Chris Rummel <crummel@microsoft.com>
Date: Fri, 4 Sep 2020 09:32:49 -0500
Subject: [PATCH] Don't restore tools in source-build

---
 eng/build.sh | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/eng/build.sh b/eng/build.sh
index eab7422de0a..cf04036b5be 100755
--- a/eng/build.sh
+++ b/eng/build.sh
@@ -319,9 +319,9 @@ if [[ "$restore" == true || "$test_core_clr" == true ]]; then
   install=true
 fi
 InitializeDotNetCli $install
-if [[ "$restore" == true ]]; then
-  dotnet tool restore
-fi
+# if [[ "$restore" == true ]]; then
+#   dotnet tool restore
+# fi
 
 bootstrap_dir=""
 if [[ "$bootstrap" == true ]]; then
-- 
2.21.3
